### DISCUSSION TITLE
xUnit parallel performance - Strange results

### DISCUSSION THREAD

**Discussion by nealrobben** [xunit/xunit#2288] [Question] [answered]
I am looking at parallelization settings in xUnit to attempt to improve the performance of unit tests at my current job.
To do this I have created a small demo project in which I experiment with the settings via xunit.runner.json files.

I am getting some strange results during my test runs and I hope someone can help me explain them

The code is written in .Net 4.5.2 with the latest version of xUnit.
The repo is located at: https://github.com/nealrobben/xUnitDemo
The unit tests are run via "dotnet test" on the command line and I measure total run time by the time displayed on my commandline prompt before and after executing the tests
(I have modified my powershell prompt to always display the time)

My solution consists of 3 projects:

DemoProject.Main: Contains 1 class with 4 methods (A/B/C/D) that each just contain Thread.Sleep(10000)
DemoProject.Main.Tests: Contains 2 classes with tests for methods A/B and method C (in 2 separate test collections)
DemoProject.Main.Tests2: Contains 1 class with a test for method D

I have added xunit.runner.json files to both test projects in which I experiment with combinations of these 2 settings:
parallelizeAssembly
parallelizeTestCollections

The setting maxParallelThreads is always set to zero.

I consistently get the following results:

1) parallelizeAssembly (false), parallelizeTestCollections (false): 35/36 seconds
2) parallelizeAssembly (false), parallelizeTestCollections (true): 24/25 seconds
3) parallelizeAssembly (true), parallelizeTestCollections (true): 24/25 seconds

Now I have 2 questions:

A) In the first case where all parallelization is disabled I would expect the duration to be at least 40 seconds since
each of my tests does a Thread.Sleep for 10 seconds. How can my time be only 35/36 seconds every time?

B) Why is there no difference between cases 2 & 3? In case 3 I enable the running of my separate test assemblies in parallel.
I would think this would also speed up execution since there is one more Thread.Sleep(10 000) that can occur in parallel in this case?

Can anyone please help me understand what I am seeing because this is all very confusing to me.
Thanks a lot!

**Comment by bradwilson** [comment, marked as answer]
I think in all cases, you're running afoul of the fact that VSTest does not obey our parallelizeAssembly request, and runs your assemblies in parallel always (there may be a VSTest-specific way to disable this, but I don't know what it is off the top of my head, assuming it exists).

Using the xUnit.net console runner, here is my result for scenario 1:

![image](https://user-images.githubusercontent.com/16944/119239431-d4865e00-bafd-11eb-8e5b-d8c5b068d0f6.png)

As you can see, 40 seconds. Exactly what we expected.

Here are my results for scenario 2:

![image](https://user-images.githubusercontent.com/16944/119239314-21b60000-bafd-11eb-877c-031ae7a948e1.png)

Note assemblies are still not being parallelized, but test collections are, so we get the expected 30 second result.

Here are my results for scenario 3:

![image](https://user-images.githubusercontent.com/16944/119239342-4b6f2700-bafd-11eb-9e76-f585e0f8f524.png)

Now with the assemblies being parallelized, we record 30 seconds of runtime but also show 20 seconds of clock time (the parenthesized number) which shows that we were running everything in parallel.
